src/datatablerow/datatablerow.jl

```diff
@@ -37,31 +35,119 @@ Base.done(r::DataTableRow, s) = s > length(r)
 
 Base.convert(::Type{Array}, r::DataTableRow) = convert(Array, r.dt[r.row,:])
 
+Base.collect(r::DataTableRow) = Tuple{Symbol, Any}[x for x in r]
+
+# the equal elements of nullable and normal arrays would have the same hashes
+const NULL_MAGIC = 0xBADDEED # what to hash if the element is null
+
+# hash column element
+Base.@propagate_inbounds hash_colel(v::AbstractArray, i, h::UInt = zero(UInt)) = hash(v[i], h)
+Base.@propagate_inbounds hash_colel{T}(v::NullableArray{T}, i, h::UInt = zero(UInt)) =
+    isnull(v, i) ? hash(NULL_MAGIC, h) : hash(get(v[i]), h)
+Base.@propagate_inbounds hash_colel{T}(v::AbstractCategoricalArray{T}, i, h::UInt = zero(UInt)) =
+    hash(CategoricalArrays.index(v.pool)[v.refs[i]], h)
```

Comment:
I guess it's not as efficient. But in these functions the constraint is to generate the make hashes invariant to the value representation: whether it's nullable or not and whether it's stored "as is" or in a categorical array. Otherwise joins would not work (we may require that joins only use the columns of identical types, but that would result in too much overhead on the user side). So we have to check if the default hash functions have this property (`Nullable` AFAIR is not).